zilliqa/src/api/zil.rs

```diff
@@ -299,19 +299,60 @@ fn get_smart_contract_state(params: Params, node: &Arc<Mutex<Node>>) -> Result<V
 
     let is_scilla = account.code.scilla_code_and_init_data().is_some();
     if is_scilla {
+        // The configuration variable contains maximum size in KiB
+        let limit = node.config.max_smart_contract_rpc_response_size * 1024;
+        let limit = if limit == 0 {
+            limit
+        } else {
+            /* This field being set to 0 corresponds to no limit */
+            u64::MAX
+        };
+
+        let mut total_size = 0;
+
         let trie = state.get_account_trie(address)?;
         for (k, v) in trie.iter() {
-            let (var_name, indices) = split_storage_key(&k)?;
-            let mut var = result.entry(var_name);
+            let (mut var_name, indices) = split_storage_key(&k)?;
+            let mut var = result.entry(var_name.clone());
+
+            // Length of a key once encoded in JSON, quotes and escaped characters included
```

Comment:
Given the complexity of calculating the length as we go, I wonder if it would be more sensible to instead limit the response size by total number of keys, rather than the serialised size? Maybe there are edge cases where values could be very large, but I'd imagine that's rare and we could always solve that with a per-value byte limit instead (which is much easier to check).

Another thought is that once we enable this limit, we will probably want to give clients a way of paginating through a contract's state which is larger than the limit. It seems more immediately natural that we would enable this pagination by letting them passing the number of keys to skip, which corresponds with a response size limited by number of entries too.

Happy to discuss though if you think otherwise? (and sorry, I think I am now contradicting an earlier opinion of mine :stuck_out_tongue: )